### DISCUSSION TITLE
ContentElement concurrency

### DISCUSSION THREAD

**Discussion by tropcicstefan** [OrchardCMS/OrchardCore#15107] [Q&A]
I think there is a bug in ContentElement, but I am not completely sure so I opened a discussion.

`ContentElement` class has `_elements` dictionary that is created on property `Elements`  and if you ask me that is not thread safe. For instance if you use `ContentExtensions.Get` in GraphQL which is implemented in parallel, or maybe in handler, you could rewrite entire dictionary and it could lead to unwanted behaviour.

What do you think?

**Comment by Piedone** [comment]
It's not thread-safe indeed. However, most APIs in Orchard Core aren't meant to be, the idea being that in a web app the multithreading you usually need to focus on is serving requests in parallel, not parallelizing within a request.

Where exactly do you see this being an issue, which GraphQL operation? (Note that `ContentElement` objects are not shared between requests.)

**Reply by tropcicstefan** [reply]
on graphql object of a custom part we have field with `ResolveAsync` where from `ResolveFieldContext.Source.ContentItem` I try to extract two parts with extension method `As<Part>()` and from time to time i get an exception 
`new InvalidCastException($"Failed casting content to '{typeof(TElement).Name}', check you have registered your content part with AddContentPart?");`
Tried with `ResolveLockedAsync` same thing

**Reply by Piedone** [reply]
This might be a bug. Could you please open a bug report in the issue tracker, with all the details you can gather?
